apiserver/authentication/agent.go

```diff
@@ -7,15 +7,55 @@ import (
 	"context"
 
 	"github.com/juju/errors"
+	"github.com/juju/names/v5"
 
 	apiservererrors "github.com/juju/juju/apiserver/errors"
 	"github.com/juju/juju/state"
 )
 
-// EntityAuthenticator performs authentication for juju entities.
-type AgentAuthenticator struct{}
+// Logger is the minimal logging interface required by the authenticator.
+type Logger interface {
+	Debugf(string, ...interface{})
+}
+
+// AgentAuthenticatorFactory is a factory for creating authenticators, which
+// can create authenticators for a given state.
+type AgentAuthenticatorFactory struct {
+	legacyState *state.State
+	logger      Logger
+}
+
+// NewAgentAuthenticatorFactory returns a new agent authenticator factory, for
+// a known state.
+func NewAgentAuthenticatorFactory(legacyState *state.State, logger Logger) AgentAuthenticatorFactory {
+	return AgentAuthenticatorFactory{
+		legacyState: legacyState,
+		logger:      logger,
+	}
+}
+
+// Authenticator returns an authenticator using the factory's state.
+func (f AgentAuthenticatorFactory) Authenticator() EntityAuthenticator {
+	return agentAuthenticator{
+		state:  f.legacyState,
+		logger: f.logger,
+	}
+}
 
-var _ EntityAuthenticator = (*AgentAuthenticator)(nil)
+// AuthenticatorForState returns an authenticator for the given state.
+func (f AgentAuthenticatorFactory) AuthenticatorForState(st *state.State) EntityAuthenticator {
+	return agentAuthenticator{
+		state:  st,
+		logger: f.logger,
+	}
+}
+
+type agentAuthenticator struct {
+	state  *state.State
+	logger Logger
+}
+
+var _ EntityAuthenticator = (*agentAuthenticator)(nil)
```

Comment:
I prefer to have these in the test files.

Reply:
I just moved them, but I'll put them in test files.